Bulk items fail due to TranslogException via ClosedChannelException

We recently made the jump to start testing ES 2.0 against our codebase, which was pretty straightforward with only a couple of changes on the Query DSL.

However, once we put this into our test environment, we see constant streams of this (please note I have removed sensitive information, such as source, index and type):

```
[2015-11-19 18:21:10,865][DEBUG][action.bulk              ] [La Nuit] [pem:2015:47][5] failed to execute bulk item (index) index {[my-index][my-type][3279b5669abaf1458583678b37bbf2a2], source[{"my-source":true}]}
[my-index][[my-index][5]] TranslogException[Failed to write operation [Index{id='3279b5669abaf1458583678b37bbf2a2', type='my-type'}]]; nested: ClosedChannelException;
	at org.elasticsearch.index.translog.Translog.add(Translog.java:497)
	at org.elasticsearch.index.engine.InternalEngine.innerIndex(InternalEngine.java:532)
	at org.elasticsearch.index.engine.InternalEngine.index(InternalEngine.java:447)
	at org.elasticsearch.index.shard.IndexShard.index(IndexShard.java:556)
	at org.elasticsearch.index.engine.Engine$Index.execute(Engine.java:815)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.executeIndexRequestOnPrimary(TransportReplicationAction.java:1073)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardIndexOperation(TransportShardBulkAction.java:338)
	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:131)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.performOnPrimary(TransportReplicationAction.java:579)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase$1.doRun(TransportReplicationAction.java:452)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.nio.channels.ClosedChannelException
	at sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:110)
	at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:199)
	at org.elasticsearch.common.io.Channels.writeToChannel(Channels.java:206)
	at org.elasticsearch.index.translog.BufferingTranslogWriter.flush(BufferingTranslogWriter.java:75)
	at org.elasticsearch.index.translog.BufferingTranslogWriter.add(BufferingTranslogWriter.java:62)
	at org.elasticsearch.index.translog.Translog.add(Translog.java:489)
	... 13 more
```

This is a pretty constant thing, and there seems to be nothing around about what might cause it. All of our ES settings are pretty much the defaults, and we haven't (deliberately) changed anything between versions. 

Is there some change in ES 2.0 which might explain the above, such as maybe tweaking default thread counts for bulk (etc)? If so, where would I start looking to resolve this exception?

If I had to guess at the rate of failure, I'd say that `> 50%` of bulk items receive this error. 

I'm not sure if this is related, but this is also showing up inside the logs; perhaps something regarding indexing is badly configured?

```
[2015-11-19 18:21:48,456][WARN ][indices.memory           ] [La Nuit] failed to set shard [my-index][4] index buffer to [309.3mb]
```

@clintongormley sure, see below. At a glance, it appears we are simply using a vanilla install (there are no cluster level settings). This is the config file our install is pointing to. I'll also point out in advance that this is currently only being used locally (i.e. from the same box), it doesn't allow outside connection.

```
# ======================== Elasticsearch Configuration =========================
#
# NOTE: Elasticsearch comes with reasonable defaults for most settings.
#       Before you set out to tweak and tune the configuration, make sure you
#       understand what are you trying to accomplish and the consequences.
#
# The primary way of configuring a node is via this file. This template lists
# the most important settings you may want to configure for a production cluster.
#
# Please see the documentation for further information on configuration options:
# <http://www.elastic.co/guide/en/elasticsearch/reference/current/setup-configuration.html>
#
# ---------------------------------- Cluster -----------------------------------
#
# Use a descriptive name for your cluster:
#

#cluster.name: elasticsearch

#
# ------------------------------------ Node ------------------------------------
#
# Use a descriptive name for the node:
#
# node.name: node-1
#
# Add custom attributes to the node:
#
# node.rack: r1
#
# ----------------------------------- Paths ------------------------------------
#
# Path to directory where to store the data (separate multiple locations by comma):
#

path.data: /mnt/data/elasticsearch/

#
# Path to log files:
#
# path.logs: /path/to/logs
#
# ----------------------------------- Memory -----------------------------------
#
# Lock the memory on startup:
#

bootstrap.mlockall: true

#
# Make sure that the `ES_HEAP_SIZE` environment variable is set to about half the memory
# available on the system and that the owner of the process is allowed to use this limit.
#
# Elasticsearch performs poorly when the system is swapping the memory.
#
# ---------------------------------- Network -----------------------------------
#
# Set the bind adress to a specific IP (IPv4 or IPv6):
#
# network.host: 192.168.0.1
#
# Set a custom port for HTTP:
#

http.port: 9200

#
# For more information, see the documentation at:
# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html>
#
# ---------------------------------- Gateway -----------------------------------
#
# Block initial recovery after a full cluster restart until N nodes are started:
#
# gateway.recover_after_nodes: 3
#
# For more information, see the documentation at:
# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-gateway.html>
#
# --------------------------------- Discovery ----------------------------------
#
# Elasticsearch nodes will find each other via unicast, by default.
#
# Pass an initial list of hosts to perform discovery when new node is started:
# The default list of hosts is ["127.0.0.1", "[::1]"]
#
# discovery.zen.ping.unicast.hosts: ["host1", "host2"]
#
# Prevent the "split brain" by configuring the majority of nodes (total number of nodes / 2 + 1):
#
# discovery.zen.minimum_master_nodes: 3
#
# For more information, see the documentation at:
# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-discovery.html>
#
# ---------------------------------- Various -----------------------------------
#
# Disable starting multiple nodes on a single system:
#
# node.max_local_storage_nodes: 1
#
# Require explicit names when deleting indices:
#
# action.destructive_requires_name: true
```

@mikemccand this is only a single node at the moment, because we're just running sanity-type stuff. Indexing rate is only a couple of events per second, grouped into bulk using the `BulkProcessor` and released in 5s batches. If we are closing the index, it's not *us* actively doing it, we've never closed indexes ;) I'm not sure exactly what's relevant from TRACE, so I grabbed a bunch of output above the actual Exception in hopes of finding the correct thing (this is before all of the Exceptions, so I assume it's useful).

```
[2015-11-19 21:07:21,232][TRACE][index.warmer             ] [La Nuit] [my-index][4] warming took [1.9ms]
[2015-11-19 21:07:21,232][TRACE][indices                  ] [La Nuit] [my-index][4] top warming [WarmerContext: ElasticsearchDirectoryReader(FilterLeafReader(_7nu(5.2.1):C44162) FilterLeafReader(_7nv(5.2.1):c6) FilterLeafReader(_7nw(5.2.1):c8) FilterLeafReader(_7nx(5.2.1):c3))]
[2015-11-19 21:07:21,232][TRACE][index.warmer             ] [La Nuit] [my-index][4] top warming took [18.4micros]
[2015-11-19 21:07:21,258][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][scheduler][T#1] DW: anyChanges? numDocsInRam=3 deletes=false hasTickets:false pendingChangesInFullFlush: false
[2015-11-19 21:07:21,258][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][scheduler][T#1] IW: nrtIsCurrent: infoVersion matches: false; DW changes: true; BD changes: false
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DW: anyChanges? numDocsInRam=3 deletes=false hasTickets:false pendingChangesInFullFlush: false
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: nrtIsCurrent: infoVersion matches: false; DW changes: true; BD changes: false
[2015-11-19 21:07:21,259][TRACE][index.shard              ] [La Nuit] [my-index][1] refresh with source: schedule
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DW: anyChanges? numDocsInRam=3 deletes=false hasTickets:false pendingChangesInFullFlush: false
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: nrtIsCurrent: infoVersion matches: false; DW changes: true; BD changes: false
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: flush at getReader
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DW: startFullFlush
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DW: anyChanges? numDocsInRam=3 deletes=false hasTickets:false pendingChangesInFullFlush: false
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DWFC: addFlushableState DocumentsWriterPerThread [pendingDeletes=gen=0, segment=_7pt, aborted=false, numDocsInRAM=3, deleteQueue=DWDQ: [ generation: 3859 ]]
[2015-11-19 21:07:21,259][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DWPT: flush postings as segment _7pt numDocs=3
[2015-11-19 21:07:21,261][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DWPT: new segment has 0 deleted docs
[2015-11-19 21:07:21,261][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DWPT: new segment has no vectors; no norms; docValues; no prox; freqs
[2015-11-19 21:07:21,261][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DWPT: flushedFiles=[_7pt_Lucene50_0.dvm, _7pt_Lucene50_0.tip, _7pt_Lucene50_0.doc, _7pt_Lucene50_0.tim, _7pt.fdx, _7pt.fdt, _7pt_Lucene50_0.dvd, _7pt.fnm]
[2015-11-19 21:07:21,261][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DWPT: flushed codec=Lucene50
[2015-11-19 21:07:21,261][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DWPT: flushed: segment=_7pt ramUsed=0.589 MB newFlushedSize=0.006 MB docs/MB=533.084
[2015-11-19 21:07:21,261][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: create compound file
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DW: publishFlushedSegment seg-private updates=null
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: publishFlushedSegment
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: publish sets newSegment delGen=7943 seg=_7pt(5.2.1):c3
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: now checkpoint "_7ps(5.2.1):C43988 _7pt(5.2.1):c3" [2 segments ; isCommit = false]
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: 0 msec to checkpoint
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: apply all deletes during flush
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: now apply all deletes for all segments maxDoc=43991
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] BD: applyDeletes: open segment readers took 0 msec
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] BD: applyDeletes: no segments; skipping
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] BD: prune sis=segments_43: _7ps(5.2.1):C43988 _7pt(5.2.1):c3 minGen=6713 packetCount=0
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: return reader version=20315 reader=StandardDirectoryReader(segments_43:20315:nrt _7ps(5.2.1):C43988 _7pt(5.2.1):c3)
[2015-11-19 21:07:21,262][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] DW: elasticsearch[La Nuit][refresh][T#9] finishFullFlush success=true
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete new file "_7pt_Lucene50_0.dvm"
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete "_7pt_Lucene50_0.dvm"
[2015-11-19 21:07:21,263][TRACE][index.store.deletes      ] [La Nuit][my-index][1] StoreDirectory.deleteFile: delete file _7pt_Lucene50_0.dvm
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete new file "_7pt_Lucene50_0.tip"
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete "_7pt_Lucene50_0.tip"
[2015-11-19 21:07:21,263][TRACE][index.store.deletes      ] [La Nuit][my-index][1] StoreDirectory.deleteFile: delete file _7pt_Lucene50_0.tip
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete new file "_7pt_Lucene50_0.doc"
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete "_7pt_Lucene50_0.doc"
[2015-11-19 21:07:21,263][TRACE][index.store.deletes      ] [La Nuit][my-index][1] StoreDirectory.deleteFile: delete file _7pt_Lucene50_0.doc
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete new file "_7pt_Lucene50_0.tim"
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete "_7pt_Lucene50_0.tim"
[2015-11-19 21:07:21,263][TRACE][index.store.deletes      ] [La Nuit][my-index][1] StoreDirectory.deleteFile: delete file _7pt_Lucene50_0.tim
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete new file "_7pt.fdx"
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete "_7pt.fdx"
[2015-11-19 21:07:21,263][TRACE][index.store.deletes      ] [La Nuit][my-index][1] StoreDirectory.deleteFile: delete file _7pt.fdx
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete new file "_7pt.fdt"
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete "_7pt.fdt"
[2015-11-19 21:07:21,263][TRACE][index.store.deletes      ] [La Nuit][my-index][1] StoreDirectory.deleteFile: delete file _7pt.fdt
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete new file "_7pt_Lucene50_0.dvd"
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete "_7pt_Lucene50_0.dvd"
[2015-11-19 21:07:21,263][TRACE][index.store.deletes      ] [La Nuit][my-index][1] StoreDirectory.deleteFile: delete file _7pt_Lucene50_0.dvd
[2015-11-19 21:07:21,263][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete new file "_7pt.fnm"
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw.ifd] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IFD: delete "_7pt.fnm"
[2015-11-19 21:07:21,264][TRACE][index.store.deletes      ] [La Nuit][my-index][1] StoreDirectory.deleteFile: delete file _7pt.fnm
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] TMP: findMerges: 2 segments
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] TMP:   seg=_7ps(5.2.1):C43988 size=7.704 MB
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] TMP:   seg=_7pt(5.2.1):c3 size=0.006 MB [floored]
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] TMP:   allowedSegmentCount=4 vs count=2 (eligible count=2) tooBigCount=0
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] MS: now merge
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] MS:   index: _7ps(5.2.1):C43988 _7pt(5.2.1):c3
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] MS:   no more merges pending; now return
[2015-11-19 21:07:21,264][TRACE][index.engine.lucene.iw   ] [La Nuit] [my-index][1] elasticsearch[La Nuit][refresh][T#9] IW: getReader took 5 msec
[2015-11-19 21:07:21,264][TRACE][indices                  ] [La Nuit] [my-index][1] warming [WarmerContext: MultiReader(FilterLeafReader(_7pt(5.2.1):c3))]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [38.3micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [36micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [72.6micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [38.3micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [39.6micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [26.8micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [29.5micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [26.1micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [38micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [24.5micros]
[2015-11-19 21:07:21,264][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [35.2micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [33.9micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [45.4micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [34.8micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [46.4micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [55.3micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [35.4micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [35.2micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [37.4micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [36micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [40.2micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [32.8micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [30.2micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [28.4micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [30.8micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [32.9micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [36.4micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [39.4micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [37.5micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [34.8micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [33.7micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [34.5micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [29.5micros]
[2015-11-19 21:07:21,265][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [39.5micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warmed bitset for [QueryWrapperFilter(+*:* -QueryWrapperFilter(_type:__*))], took [24.8micros]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] warming took [2.2ms]
[2015-11-19 21:07:21,266][TRACE][indices                  ] [La Nuit] [my-index][1] top warming [WarmerContext: ElasticsearchDirectoryReader(FilterLeafReader(_7ps(5.2.1):C43988) FilterLeafReader(_7pt(5.2.1):c3))]
[2015-11-19 21:07:21,266][TRACE][index.warmer             ] [La Nuit] [my-index][1] top warming took [18micros]
[2015-11-19 21:07:21,742][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [16440] bytes, now [16440]
[2015-11-19 21:07:21,742][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [-16440] bytes, now [0]
[2015-11-19 21:07:23,639][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [16440] bytes, now [16440]
[2015-11-19 21:07:23,639][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [-16440] bytes, now [0]
[2015-11-19 21:07:24,839][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [16440] bytes, now [16440]
[2015-11-19 21:07:24,839][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [-16440] bytes, now [0]
[2015-11-19 21:07:26,433][TRACE][marvel.agent             ] [La Nuit] collecting data - collectors [index-stats-collector,cluster-stats-collector,cluster-info-collector,cluster-state-collector,index-recovery-collector,indices-stats-collector,shards-collector,node-stats-collector]
[2015-11-19 21:07:26,433][TRACE][marvel.agent.collector.indices] [La Nuit] collector [index-stats-collector] - collecting data...
[2015-11-19 21:07:26,508][TRACE][transport.tracer         ] [La Nuit] [5024019][indices:monitor/stats[n]] received response from [{La Nuit}{RPKw6aGPR-ilRvsQtufC4g}{127.0.0.1}{127.0.0.1:9300}]
[2015-11-19 21:07:26,508][TRACE][marvel.agent             ] [La Nuit] bulk [all] - adding [35] collected docs from [index-stats-collector] collector
[2015-11-19 21:07:26,510][TRACE][marvel.agent.collector.cluster] [La Nuit] collector [cluster-stats-collector] - collecting data...
[2015-11-19 21:07:26,517][TRACE][transport.tracer         ] [La Nuit] [2227][indices:data/write/bulk] received request
[2015-11-19 21:07:26,518][TRACE][index.shard              ] [La Nuit] [my-index][3] index [my-type][e942d5a4692df8f10d6d657067eb6d30][org.elasticsearch.index.mapper.ParseContext$Document@56eef3e5]
[2015-11-19 21:07:26,518][TRACE][index.shard              ] [La Nuit] [my-index][4] index [my-type][b661569def0e70e29ce28ff8c9933e44][org.elasticsearch.index.mapper.ParseContext$Document@7f97cfc5]
[2015-11-19 21:07:26,518][TRACE][index.shard              ] [La Nuit] [my-index][6] index [my-type][d64ba78b781f63f1c5c6fe590131601d][org.elasticsearch.index.mapper.ParseContext$Document@5d6fdfda]
[2015-11-19 21:07:26,518][TRACE][index.shard              ] [La Nuit] [my-index][7] index [my-type][60c150a56289b0f815bbe840165fe6d0][org.elasticsearch.index.mapper.ParseContext$Document@75a1fcab, org.elasticsearch.index.mapper.ParseContext$Document@4bdda35b]
[2015-11-19 21:07:26,518][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [16440] bytes, now [16440]
[2015-11-19 21:07:26,518][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [16440] bytes, now [32880]
[2015-11-19 21:07:26,518][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [-16440] bytes, now [16440]
[2015-11-19 21:07:26,518][TRACE][indices.breaker          ] [La Nuit] [request] Adjusted breaker by [-16440] bytes, now [0]
[2015-11-19 21:07:26,518][DEBUG][action.bulk              ] [La Nuit] [my-index][3] failed to execute bulk item (index) index {[my-index][my-type][e942d5a4692df8f10d6d657067eb6d30], source[{"my-source":true}]}
[my-index][[my-index][3]] TranslogException[Failed to write operation [Index{id='e942d5a4692df8f10d6d657067eb6d30', type='my-type'}]]; nested: ClosedChannelException;
    at org.elasticsearch.index.translog.Translog.add(Translog.java:497)
    at org.elasticsearch.index.engine.InternalEngine.innerIndex(InternalEngine.java:532)
    at org.elasticsearch.index.engine.InternalEngine.index(InternalEngine.java:447)
    at org.elasticsearch.index.shard.IndexShard.index(IndexShard.java:556)
    at org.elasticsearch.index.engine.Engine$Index.execute(Engine.java:815)
    at org.elasticsearch.action.support.replication.TransportReplicationAction.executeIndexRequestOnPrimary(TransportReplicationAction.java:1073)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.shardIndexOperation(TransportShardBulkAction.java:338)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:131)
    at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase.performOnPrimary(TransportReplicationAction.java:579)
    at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryPhase$1.doRun(TransportReplicationAction.java:452)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.nio.channels.ClosedChannelException
    at sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:110)
    at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:199)
    at org.elasticsearch.common.io.Channels.writeToChannel(Channels.java:206)
    at org.elasticsearch.index.translog.BufferingTranslogWriter.flush(BufferingTranslogWriter.java:75)
    at org.elasticsearch.index.translog.BufferingTranslogWriter.add(BufferingTranslogWriter.java:62)
    at org.elasticsearch.index.translog.Translog.add(Translog.java:489)
    ... 13 more
```

--- a/core/src/main/java/org/elasticsearch/index/translog/BufferingTranslogWriter.java
+++ b/core/src/main/java/org/elasticsearch/index/translog/BufferingTranslogWriter.java
@@ -46,6 +46,7 @@ public final class BufferingTranslogWriter extends TranslogWriter {
 
     @Override
     public Translog.Location add(BytesReference data) throws IOException {
+        ensureOpen();
         try (ReleasableLock lock = writeLock.acquire()) {
             operationCounter++;
             final long offset = totalOffset;

--- a/core/src/main/java/org/elasticsearch/index/translog/BufferingTranslogWriter.java
+++ b/core/src/main/java/org/elasticsearch/index/translog/BufferingTranslogWriter.java
@@ -46,8 +46,8 @@ public final class BufferingTranslogWriter extends TranslogWriter {
 
     @Override
     public Translog.Location add(BytesReference data) throws IOException {
-        ensureOpen();
         try (ReleasableLock lock = writeLock.acquire()) {
+            ensureOpen();
             operationCounter++;
             final long offset = totalOffset;
             if (data.length() >= buffer.length) {
@@ -107,19 +107,25 @@ public final class BufferingTranslogWriter extends TranslogWriter {
             return;
         }
         synchronized (this) {
-            try (ReleasableLock lock = writeLock.acquire()) {
-                flush();
-                lastSyncedOffset = totalOffset;
+            channelReference.incRef();
+            try {
+                try (ReleasableLock lock = writeLock.acquire()) {
+                    flush();
+                    lastSyncedOffset = totalOffset;
+                }
+                // we can do this outside of the write lock but we have to protect from
+                // concurrent syncs
+                checkpoint(lastSyncedOffset, operationCounter, channelReference);
+            } finally {
+                channelReference.decRef();
             }
-            // we can do this outside of the write lock but we have to protect from
-            // concurrent syncs
-            checkpoint(lastSyncedOffset, operationCounter, channelReference);
         }
     }
 
 
     public void updateBufferSize(int bufferSize) {
         try (ReleasableLock lock = writeLock.acquire()) {
+            ensureOpen();
             if (this.buffer.length != bufferSize) {
                 flush();
                 this.buffer = new byte[bufferSize];

--- a/core/src/main/java/org/elasticsearch/index/translog/Translog.java
+++ b/core/src/main/java/org/elasticsearch/index/translog/Translog.java
@@ -407,6 +407,7 @@ public class Translog extends AbstractIndexShardComponent implements IndexShardC
             out.seek(end);
             final ReleasablePagedBytesReference bytes = out.bytes();
             try (ReleasableLock lock = readLock.acquire()) {
+                ensureOpen();
                 Location location = current.add(bytes);
                 if (config.isSyncOnEachOperation()) {
                     current.sync();
@@ -414,6 +415,8 @@ public class Translog extends AbstractIndexShardComponent implements IndexShardC
                 assert current.assertBytesAtLocation(location, bytes);
                 return location;
             }
+        } catch (AlreadyClosedException ex) {
+            throw ex;
         } catch (Throwable e) {
             throw new TranslogException(shardId, "Failed to write operation [" + operation + "]", e);
         } finally {

--- a/core/src/main/java/org/elasticsearch/index/translog/Translog.java
+++ b/core/src/main/java/org/elasticsearch/index/translog/Translog.java
@@ -1288,8 +1288,8 @@ public class Translog extends AbstractIndexShardComponent implements IndexShardC
 
     @Override
     public void prepareCommit() throws IOException {
-        ensureOpen();
         try (ReleasableLock lock = writeLock.acquire()) {
+            ensureOpen();
             if (currentCommittingTranslog != null) {
                 throw new IllegalStateException("already committing a translog with generation: " + currentCommittingTranslog.getGeneration());
             }
@@ -1321,9 +1321,9 @@ public class Translog extends AbstractIndexShardComponent implements IndexShardC
 
     @Override
     public void commit() throws IOException {
-        ensureOpen();
         ImmutableTranslogReader toClose = null;
         try (ReleasableLock lock = writeLock.acquire()) {
+            ensureOpen();
             if (currentCommittingTranslog == null) {
                 prepareCommit();
             }

--- a/core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java
+++ b/core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java
@@ -123,9 +123,9 @@ public class TranslogWriter extends TranslogReader {
      * add the given bytes to the translog and return the location they were written at
      */
     public Translog.Location add(BytesReference data) throws IOException {
-        ensureOpen();
         final long position;
         try (ReleasableLock lock = writeLock.acquire()) {
+            ensureOpen();
             position = writtenOffset;
             data.writeTo(channel);
             writtenOffset = writtenOffset + data.length();
@@ -200,9 +200,9 @@ public class TranslogWriter extends TranslogReader {
      * returns a new immutable reader which only exposes the current written operation *
      */
     public ImmutableTranslogReader immutableReader() throws TranslogException {
-        ensureOpen();
         if (channelReference.tryIncRef()) {
             try (ReleasableLock lock = writeLock.acquire()) {
+                ensureOpen();
                 flush();
                 ImmutableTranslogReader reader = new ImmutableTranslogReader(this.generation, channelReference, firstOperationOffset, writtenOffset, operationCounter);
                 channelReference.incRef(); // for new reader

--- a/core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java
+++ b/core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java
@@ -460,36 +460,7 @@ public class TranslogTests extends ESTestCase {
         final CountDownLatch downLatch = new CountDownLatch(1);
         for (int i = 0; i < threadCount; i++) {
             final int threadId = i;
-            threads[i] = new Thread(new Runnable() {
-                @Override
-                public void run() {
-                    try {
-                        downLatch.await();
-                        for (int opCount = 0; opCount < opsPerThread; opCount++) {
-                            Translog.Operation op;
-                            switch (randomFrom(Translog.Operation.Type.values())) {
-                                case CREATE:
-                                case INDEX:
-                                    op = new Translog.Index("test", threadId + "_" + opCount,
-                                            randomUnicodeOfLengthBetween(1, 20 * 1024).getBytes("UTF-8"));
-                                    break;
-                                case DELETE:
-                                    op = new Translog.Delete(new Term("_uid", threadId + "_" + opCount),
-                                            1 + randomInt(100000),
-                                            randomFrom(VersionType.values()));
-                                    break;
-                                default:
-                                    throw new ElasticsearchException("not supported op type");
-                            }
-
-                            Translog.Location loc = translog.add(op);
-                            writtenOperations.add(new LocationOperation(op, loc));
-                        }
-                    } catch (Throwable t) {
-                        threadExceptions[threadId] = t;
-                    }
-                }
-            });
+            threads[i] = new TranslogThread(translog, downLatch, opsPerThread, threadId, writtenOperations, threadExceptions);
             threads[i].setDaemon(true);
             threads[i].start();
         }
@@ -1218,6 +1189,94 @@ public class TranslogTests extends ESTestCase {
                 assertEquals(Integer.parseInt(next.getSource().source.toUtf8()), i);
             }
             assertNull(snapshot.next());
+        }
+    }
+
+    public void testFailOnClosedWrite() throws IOException {
+        translog.add(new Translog.Index("test", "1", Integer.toString(1).getBytes(Charset.forName("UTF-8"))));
+        translog.close();
+        try {
+            translog.add(new Translog.Index("test", "1", Integer.toString(1).getBytes(Charset.forName("UTF-8"))));
+            fail("closed");
+        } catch (AlreadyClosedException ex) {
+            // all is welll
+        }
+    }
+
+    public void testCloseConcurrently() throws Throwable {
+        final int opsPerThread = randomIntBetween(10, 200);
+        int threadCount = 2 + randomInt(5);
+
+        logger.info("testing with [{}] threads, each doing [{}] ops", threadCount, opsPerThread);
+        final BlockingQueue<LocationOperation> writtenOperations = new ArrayBlockingQueue<>(threadCount * opsPerThread);
+
+        Thread[] threads = new Thread[threadCount];
+        final Throwable[] threadExceptions = new Throwable[threadCount];
+        final CountDownLatch downLatch = new CountDownLatch(1);
+        for (int i = 0; i < threadCount; i++) {
+            final int threadId = i;
+            threads[i] = new TranslogThread(translog, downLatch, opsPerThread, threadId, writtenOperations, threadExceptions);
+            threads[i].setDaemon(true);
+            threads[i].start();
+        }
+
+        downLatch.countDown();
+        translog.close();
+
+        for (int i = 0; i < threadCount; i++) {
+            if (threadExceptions[i] != null) {
+                if ((threadExceptions[i] instanceof AlreadyClosedException) == false) {
+                    throw threadExceptions[i];
+                }
+            }
+            threads[i].join(60 * 1000);
+        }
+    }
+
+    private static class TranslogThread extends Thread {
+        private final CountDownLatch downLatch;
+        private final int opsPerThread;
+        private final int threadId;
+        private final BlockingQueue<LocationOperation> writtenOperations;
+        private final Throwable[] threadExceptions;
+        private final Translog translog;
+
+        public TranslogThread(Translog translog, CountDownLatch downLatch, int opsPerThread, int threadId, BlockingQueue<LocationOperation> writtenOperations, Throwable[] threadExceptions) {
+            this.translog = translog;
+            this.downLatch = downLatch;
+            this.opsPerThread = opsPerThread;
+            this.threadId = threadId;
+            this.writtenOperations = writtenOperations;
+            this.threadExceptions = threadExceptions;
+        }
+
+        @Override
+        public void run() {
+            try {
+                downLatch.await();
+                for (int opCount = 0; opCount < opsPerThread; opCount++) {
+                    Translog.Operation op;
+                    switch (randomFrom(Translog.Operation.Type.values())) {
+                        case CREATE:
+                        case INDEX:
+                            op = new Translog.Index("test", threadId + "_" + opCount,
+                                    randomUnicodeOfLengthBetween(1, 20 * 1024).getBytes("UTF-8"));
+                            break;
+                        case DELETE:
+                            op = new Translog.Delete(new Term("_uid", threadId + "_" + opCount),
+                                    1 + randomInt(100000),
+                                    randomFrom(VersionType.values()));
+                            break;
+                        default:
+                            throw new ElasticsearchException("not supported op type");
+                    }
+
+                    Translog.Location loc = translog.add(op);
+                    writtenOperations.add(new LocationOperation(op, loc));
+                }
+            } catch (Throwable t) {
+                threadExceptions[threadId] = t;
+            }
         }
     }
 }

